File: BE-1/CL-1/DAA/TravellingSalesman/tsdynamic.py
class TSDynamic():
	def __init__(self,graph):
		self.graph=graph
		self.no=len(graph)
		self.nodes = {i for i in range(0,self.no)}
		self.costs = dict()

	def findPath(self,start):
		return self.cost(start,self.nodes-{start},start)

	def cost(self,start,via,end):
		tup = (start,tuple(i for i in via),end)
		if not tup in self.costs:
			minc = 99999;
			mintup = ()
			for i in via:
				temp = self.cost(i,via-{i},end)
				mincost = (self.graph[start][i]+temp[0] ,[start]+[j for j in temp[1]])
				if mincost[0] < minc:
					minc = mincost[0]
					mintup = mincost
			if len(via) == 0:
				self.costs[tup] = (self.graph[start][end],[start,end])
			else:
				self.costs[tup] = mintup
		return self.costs[tup]

if __name__=='__main__':
	nodes = int(input("Enter no of nodes :"))
	# graph = [[0,24,11,10,9],
	#  		[8,0,2,5,11],
	#  		[26,12,0,8,7],
	#  		[11,23,24,0,6],
	#  		[5,4,8,11,0]]
	graph=[[ 0 for i in range(nodes) ] for j in range(nodes)]
	print("Enter Edges")	
	for i in range(nodes):
		for j in range(nodes):
			if i!=j:
				graph[i][j] = input("Edge Length {} -> {} :".format(i,j))
	print("Given Problem\n----------------------------------")
	for i in range(nodes):
		for j in range(nodes):
			if i!=j:
				print(graph[i][j],end="\t")
			else:
				print("-",end="\t")
		print("\n----------------------------------")
	problem = TSDynamic(graph)
	path = problem.findPath(0)
	print("Minimum Cost : {}".format(path[0]) )
	print("Optimal Path : {}".format("->".join(list(map(lambda i : str(i+1) , path[1])))))

File: BE-1/CL-1/DAA/TravellingSalesman/tsdynamic.cpp
#include <iostream>
#include <vector>
#include <set>
#include <iterator>
#include <omp.h>
using namespace std;

class Cost{
	public:
		int start;
		set<int> nodes;
		int end;
		int cost;
		vector<int> path;
		Cost(int start,set<int> nodes,int end){
			this->start=start;
			this->nodes=nodes;
			this->end=end;
		}
		bool compare(int s,set<int> is,int e){
			if(s==start && end==e && is==nodes ) return true;
			else return false;
		}

};

class TSDynamic{
	private:
		int** graph;
		int no;
		set<int> nodes;
		vector<Cost*> costs;
	public:
		TSDynamic(int** graph,int no){
			this->graph=graph;
			this->no=no;
			for(int i=0;i<no;i++){
				this->nodes.insert(i);
			}
		}

		Cost* findPath(int start){
			return cost(start,giveSet(this->nodes,start),start);
		}

		Cost* cost(int start,set<int> via,int end){
			Cost* c = isCostThere(start,via,end);
			if(c==NULL){
				c = new Cost(start,giveSet(via,start),end);
				int minc = 9999;
				Cost* minCost;
				set<int>::iterator it;
				
				for(it=via.begin();it!=via.end();++it){
					Cost* ctemp = cost(*it,giveSet(via,*it),end);
					int ccost = graph[start][*it] + ctemp->cost;
					if(ccost<minc){
						minc=ccost;
						minCost = ctemp;
					}
				}
				if(via.size()==0){
					c->cost = graph[start][end];
					c->path.push_back(start);
					c->path.push_back(end); 
				}
				else{
					c->cost= minc;
					c->path.push_back(start);
					c->path.insert(c->path.end(),minCost->path.begin(),minCost->path.end());
				}
			}
			return c;	
		}

		set<int> giveSet(set<int> prev,int discard){
			set<int> newSet;
			set<int>::iterator it;
			for(it=prev.begin();it!=prev.end();++it){
				if(*it != discard)
					newSet.insert(*it);
			}
			return newSet;
		}

		Cost* isCostThere(int start,set<int> via,int end){
			int len = costs.size();
			Cost* ans = NULL;
			#pragma omp parallel for
			for(int i=0;i<len;i++){
				Cost* c = costs[i];
				if(c->compare(start,via,end)){
					ans=c;
				}
			}
			return ans;
		}
};

int main(){
	int nodes;cout<<"Number of Nodes :";cin>>nodes;
	int** graph = new int*[nodes];
	for(int i=0;i<nodes;i++){
		graph[i]= new int[nodes];
	}

	cout<<"Enter Problem Matrix (seperate rows by linebreak and col by blankspace)"<<endl;
	for(int i =0;i<nodes;i++){
		for(int j=0;j<nodes;j++){
			cin>>graph[i][j];
		}
	}

	TSDynamic* problem =  new TSDynamic(graph,nodes);
	double start = omp_get_wtime(); 
	Cost* sol = problem->findPath(0);
	double end = omp_get_wtime();
	cout<<"Time taken for execution :"<<end-start<<endl;
	cout<<"Minimum Cost :"<<sol->cost<<endl;
	cout<<"Optimal Path :";cout<<sol->path[0]+1;
	for(int i=1;i<sol->path.size();i++){
		cout<<"->"<<sol->path[i]+1;
	}

	cout<<endl;
	return 1;
}